[FEAT] Add support for Enterprise Teams

vmvarela opened the pull request:
Add support for managing GitHub Enterprise Teams (enterprise-level teams):

Resources:
- github_enterprise_team: create/update/delete enterprise teams
- github_enterprise_team_membership: manage team members
- github_enterprise_team_organizations: assign teams to organizations

Data sources:
- github_enterprise_team: lookup by slug or ID
- github_enterprise_teams: list all enterprise teams
- github_enterprise_team_membership: check user membership
- github_enterprise_team_organizations: list assigned orgs

Note: These endpoints require GitHub Enterprise Cloud with a classic PAT that has enterprise admin permissions.

<!-- Please refer to our contributing docs for any questions on submitting a pull request -->
<!-- Issues are required for both bug fixes and features. -->
Resolves #3001

----

### Before the change?
<!-- Please describe the current behavior that you are modifying. -->

* No support for enterprise teams

### After the change?
<!-- Please describe the behavior or changes that are being added by this PR. -->

* Added support for managing enterprise teams, memberships and assingments.

### Pull request checklist
- [ ] Schema migrations have been created if needed ([example](https://github.com/integrations/terraform-provider-github/pull/2820/files))
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?
<!-- If this introduces a breaking change make sure to note it here any what the impact might be -->

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes
- [x] No

----

deiga commented on github/data_source_github_enterprise_team.go:
```diff
@@ -0,0 +1,135 @@
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"slug"},
+				Description:   "The numeric ID of the enterprise team.",
```
I'm not sure how validation acts on Optional & Computed values, but if it works like I think it should, then this should remove the need to check for `team_id != 0` in the CRUD functions
```suggestion
				Description:   "The numeric ID of the enterprise team.",
				ValidateDiagFunc: validation.ToDiagFunc(validation.IntAtLeast(1))
```

deiga commented on github/data_source_github_enterprise_team.go:
```diff
@@ -0,0 +1,135 @@
+				Type:          schema.TypeString,
+				Optional:      true,
+				Computed:      true,
+				ConflictsWith: []string{"team_id"},
```
I think if you use `ExactlyOneOf` here and in `team_id` you can rely on either one being set at all times
```suggestion
				ExactlyOneOf: []string{"team_id"},
```

deiga reviewed (commented):
Thanks for your contribution!

The Enteprise Teams functionality isn't yet published in the `go-github` SDK, but it will most likely be coming in v81. I would suggest to refactor this PR to use the upcoming release of the SDK instead of making our own API interactions.

We most likely won't get to merging this before we release v7 of the provider, where we'll hopefully quickly get to latest SDK version.

A few other notes:
- Try to leverage `ValidateDiagFunc` in the Schema as much as possible, instead of failing in the CRUD provider functions on missing information. For more complex validations using a `CustomizeDiff` function is a good option as well. That will allow the user to get errors on `plan` instead of `apply`
- Please add top-level `Description` fields to all Schemata and ensure that all `Description` fields contain same information as the docs pages. We'll be moving to generated docs pages in the near future

deiga commented on github/data_source_github_enterprise_team.go:
```diff
@@ -0,0 +1,158 @@
+func dataSourceGithubEnterpriseTeam() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceGithubEnterpriseTeamRead,
+
```
please add a top level `Description` field as well

deiga reviewed (commented):
Half-way though. Looking good!

vmvarela commented:
> Half-way though. Looking good!

Thanks for your support, @deiga .

deiga commented on github/resource_github_enterprise_team_test.go:
```diff
@@ -0,0 +1,183 @@
+
+	t.Run("with an enterprise account", func(t *testing.T) {
+		testCase(t)
+	})
```
We no longer user the `testCase` pattern which we had previously :)
```suggestion
	resource.Test(t, resource.TestCase{
		PreCheck:  func() { skipUnlessEnterprise(t) },
		ProviderFactories: providerFactories,
		Steps: []resource.TestStep{
			{Config: config, Check: check},
			{
				ResourceName:      "github_enterprise_team_membership.test",
				ImportState:       true,
				ImportStateVerify: true,
			},
		},
	})
```

deiga replied:
Please remove it from other parts as well :)

deiga commented on github/resource_github_enterprise_team_organizations.go:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	d.SetId(buildTwoPartID(enterpriseSlug, team.Slug))
+	return resourceGithubEnterpriseTeamOrganizationsRead(ctx, d, meta)
```
See https://github.com/integrations/terraform-provider-github/pull/3008/changes#r2677759068

deiga commented on github/resource_github_enterprise_team_membership.go:
```diff
@@ -0,0 +1,130 @@
+				ForceNew:         true,
+				Description:      "The slug or ID of the enterprise team.",
+				ValidateDiagFunc: validation.ToDiagFunc(validation.All(validation.StringIsNotWhiteSpace, validation.StringIsNotEmpty)),
+			},
```
See https://github.com/integrations/terraform-provider-github/pull/3008/changes#r2677765600

github-actions[bot] committed:
feat(enterprise-teams): add utility functions for enterprise teams

Add util_enterprise_teams.go with helper functions for managing
enterprise team resources.

github-actions[bot] committed:
feat(enterprise-teams): add github_enterprise_team resource

Manages GitHub Enterprise team entities (create, read, update, delete).

Includes:
- Resource implementation with CRUD operations
- Acceptance tests
- Documentation

github-actions[bot] committed:
feat(enterprise-teams): add github_enterprise_team_membership resource

Manages membership of users in GitHub Enterprise teams.

Includes:
- Resource implementation
- Documentation

github-actions[bot] committed:
feat(enterprise-teams): add github_enterprise_team_organizations resource

Manages organization assignments to GitHub Enterprise teams.

Includes:
- Resource implementation
- Documentation

github-actions[bot] committed:
feat(enterprise-teams): add data sources for enterprise teams

Add data sources:
- github_enterprise_team: retrieve a team by slug or ID
- github_enterprise_teams: list all teams in an enterprise
- github_enterprise_team_membership: get team members
- github_enterprise_team_organizations: get team organizations

Includes:
- Data source implementations
- Acceptance tests
- Documentation

github-actions[bot] committed:
feat(enterprise-teams): register enterprise team resources and data sources

- Register 3 resources in provider.go:
  - github_enterprise_team
  - github_enterprise_team_membership
  - github_enterprise_team_organizations
- Register 4 data sources in provider.go:
  - github_enterprise_team
  - github_enterprise_teams
  - github_enterprise_team_membership
  - github_enterprise_team_organizations

github-actions[bot] committed:
fix(enterprise-teams): update go-github import to v82

Update import paths from go-github/v81 to go-github/v82 to match
the current version in upstream/main.

vmvarela commented:
**Commit history reorganized** 🔄

Hi @deiga, I've reorganized the commit history as requested. The changes are now split into logical, granular commits:

1. `feat(enterprise-teams): add utility functions for enterprise teams`
2. `feat(enterprise-teams): add github_enterprise_team resource` (+ test + docs)
3. `feat(enterprise-teams): add github_enterprise_team_membership resource` (+ docs)
4. `feat(enterprise-teams): add github_enterprise_team_organizations resource` (+ docs)
5. `feat(enterprise-teams): add data sources for enterprise teams` (4 data sources + tests + docs)
6. `feat(enterprise-teams): register enterprise team resources and data sources`
7. `fix(enterprise-teams): update go-github import to v82`

Each commit is self-contained and can be reviewed independently. Going forward, I'll create new commits for any review feedback instead of amending.

Sorry for the confusion with the previous workflow - I appreciate your patience and feedback on improving my contribution process.

vmvarela committed:
test(enterprise-teams): migrate tests to terraform-plugin-testing and fix import verification

vmvarela committed:
fix(enterprise-teams): address review findings

- Rename single-char variables: t→team, v→item (Rule #29)
- Remove obvious comments that restate the code (Rule #30)
- Fix 404 handling with errors.As instead of resp.StatusCode (Rule #13)
- Migrate all tests from Check: to ConfigStateChecks: (Rule #31)
- Fix docs verb forms: Get→Gets, List→Lists, Check→Checks (Rule #2)
- Move listAllEnterpriseTeams to util_enterprise_teams.go for consistency
- Add sidebar entries for all 7 resources/data sources
- Add unit tests for utility functions (Rule #33)

vmvarela committed:
fix(enterprise-teams): update go-github import from v82 to v83

vmvarela committed:
fix(enterprise-teams): authoritative Create/Delete for team organizations

- Create validates no pre-existing org assignments before adding
- Delete fetches all current orgs from API instead of Terraform state

Poltergeisen commented:
Looking forward to this one. I'd love to be able to manage teams with terraform. My apologies -- I dont know Go. Is this going to work with teams linked to an IdP like entra?

vmvarela committed:
fix: update go-github import to v84 after rebase

vmvarela committed:
fix(enterprise-teams): replace "Github" with "GitHub" in page_title

Fixes #22

- website/docs/d/enterprise.html.markdown
- website/docs/r/enterprise_organization.html.markdown

vmvarela committed:
fix(enterprise-teams): fix incomplete error handling in listAllEnterpriseTeamOrganizations callers

Closes #29

- Read: handle 404 by clearing state (d.SetId("")) instead of returning error,
  so resources deleted out-of-band are cleanly removed from state.
- Create: handle 404 from pre-existence check as empty assignments and proceed,
  since a new team may not have any assignments yet.
- Delete: already handled correctly — no change needed.

The errors.As check now branches differently in each caller, making the
type assertion effective as the reviewer requested.

vmvarela committed:
fix(enterprise-teams): replace "Github" with "GitHub" in page_title (#39)

Fixes #22

- website/docs/d/enterprise.html.markdown
- website/docs/r/enterprise_organization.html.markdown

vmvarela committed:
fix(enterprise-teams): fix incomplete error handling in listAllEnterpriseTeamOrganizations callers (#40)

Closes #29

- Read: handle 404 by clearing state (d.SetId("")) instead of returning error,
  so resources deleted out-of-band are cleanly removed from state.
- Create: handle 404 from pre-existence check as empty assignments and proceed,
  since a new team may not have any assignments yet.
- Delete: already handled correctly — no change needed.

The errors.As check now branches differently in each caller, making the
type assertion effective as the reviewer requested.

vmvarela committed:
fix(enterprise-teams): move input validation to schema ValidateDiagFunc

Remove redundant runtime checks from dataSourceGithubEnterpriseTeamRead:
- Drop 'if teamID != 0' guard — ValidateDiagFunc(IntAtLeast(1)) already
  rejects zero/negative values at plan time.
- Drop 'if teamSlug == ""' fallback — ExactlyOneOf already enforces that
  exactly one of slug or team_id is provided.

Closes #31

vmvarela committed:
fix(enterprise-teams): fix grammar in resource Description fields

Use "Creates and manages..." (third-person singular) in all three
enterprise-teams resources, consistent with provider conventions and
reviewer feedback on upstream PR #3008.

Closes #23

vmvarela committed:
fix(enterprise-teams): use read-only language in data source Descriptions

Replace "Lists..." with "Retrieves information about..." in:
- data_source_github_enterprise_teams.go
- data_source_github_enterprise_team_organizations.go

Data sources are read-only; their Description must not imply mutation.
Consistent with provider conventions and reviewer feedback on upstream
PR #3008.

Closes #24

vmvarela committed:
fix(teams): omit group_id from Create/Update payload when empty

Only set GroupID in the API request when non-empty, consistent with
how description is handled. Sending an empty string pointer serializes
as "group_id": "" in the JSON payload which may cause unexpected
API behavior. Omitting the field (nil pointer with omitempty) is the
correct way to create a team without a group assignment.

Closes #56

vmvarela committed:
chore(lint): exclude modernize newexpr rule for Ptr calls

The modernize/newexpr rule incorrectly suggests replacing github.Ptr(x)
with new(x), but new() takes a type not an expression — the auto-fix
generates invalid Go code. Exclude this specific rule pattern.

bollicino commented:
ciao @vmvarela thanks for the great effort on this feature: enterprise teams management in Terraform is a huge addition! 
I wanted to check: is there a plan to merge this soon? If so, when can we expect it to land in a release?

Appreciate the work! ❤️

jmaixl commented:
+1, now that enterprise teams are GA, will this be re-opened (closed for some reason)?

muzfuz commented:
Appreciate all the work that's gone into this. It would definitely help us a lot as well.

Paul-Weaver commented:
Hey @vmvarela, any update here? Now that Enterprise Teams are GA, we’d love to know if there are plans to reopen/merge this, as it would unblock us from managing enterprise teams through Terraform.
